python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -752,17 +730,40 @@ def prepare(self, containers):
 class PowerSupplyOrFan(Shadow):
     __shadowclass__ = manage.PowerSupplyOrFan
     __lookups__ = [('netbox', 'name')]
+    fan_event = EventFactory('ipdevpoll', 'eventEngine', 'fanState')
+    psu_event = EventFactory('ipdevpoll', 'eventEngine', 'psuState')
 
     def prepare(self, containers):
-        existing = self.get_existing_model(containers)
+        self.is_new = not self.get_existing_model(containers)
         # Set a default value of UNKNOWN if this is a new object
-        if not existing and self.up is None:
+        if self.is_new and self.up is None:
             self.up = manage.PowerSupplyOrFan.STATE_UNKNOWN
 
     @classmethod
     def cleanup_after_save(cls, containers):
+        cls._handle_new_psus_and_fans(containers)
         cls._delete_missing_psus_and_fans(containers)
 
+    @classmethod
+    def _handle_new_psus_and_fans(cls, containers):
+        for shadow_psufan in containers[cls].values():
+            if shadow_psufan.is_new:
+                psufan = shadow_psufan.get_existing_model()
+                if psufan.is_psu():
+                    cls.psu_event.notify(
+                        device=psufan.device,
+                        netbox=psufan.netbox,
+                        subid=psufan.id,
+                        alert_type="psuNew",
+                    ).save()
+                elif psufan.is_fan():
+                    cls.fan_event.notify(
+                        device=psufan.device,
+                        netbox=psufan.netbox,
+                        subid=psufan.id,
+                        alert_type="fanNew",
+                    ).save()
```

Comment:
Suggesting a more compact version, without the redundancies:

```suggestion
                cls.psu_event.notify(
                    device=psufan.device,
                    netbox=psufan.netbox,
                    subid=psufan.id,
                    alert_type="fanNew" if psufan.is_fan() else "psuNew",
                ).save()
```